bpcosmo/pm.py

```diff
@@ -0,0 +1,144 @@
+# This module contains utility functions to run a forward simulation
+import jax
+import jax.numpy as jnp
+
+import jax_cosmo as jc
+
+import numpyro
+import numpyro.distributions as dist
+
+from jax.experimental.ode import odeint
+from jaxpm.pm import lpt, make_ode_fn
+from jaxpm.kernels import fftk
+from jaxpm.lensing import density_plane
+import haiku as hk
+
+from jaxpm.painting import cic_paint, cic_read
+from jaxpm.kernels import gradient_kernel, laplace_kernel, longrange_kernel
+from jaxpm.nn import NeuralSplineFourierFilter
+
+__all__ = [
+    'get_density_planes',
+]
+
+model = hk.without_apply_rng(hk.transform(lambda x,a : NeuralSplineFourierFilter(n_knots=16, latent_size=32)(x,a)))
+
+def linear_field(mesh_shape, box_size, pk):
+    """
+    Generate initial conditions.
+    """
+    kvec = fftk(mesh_shape)
+    kmesh = sum((kk / box_size[i] * mesh_shape[i])**2 for i, kk in enumerate(kvec))**0.5
+    pkmesh = pk(kmesh) * (mesh_shape[0] * mesh_shape[1] * mesh_shape[2]) / (box_size[0] * box_size[1] * box_size[2])
+
+    field = numpyro.sample('initial_conditions',
+                           dist.Normal(jnp.zeros(mesh_shape), 
+                                       jnp.ones(mesh_shape)))
+
+    field = jnp.fft.rfftn(field) * pkmesh**0.5
+    field = jnp.fft.irfftn(field)
+    return field
+
+def get_density_planes(cosmology,
+                   density_plane_width=100.,       # In Mpc/h
+                   density_plane_npix=256,         # Number of pixels
+                   density_plane_smoothing=3.,     # In Mpc/h
+                   box_size=[400.,400.,4000.],     # In Mpc/h
+                   nc=[64,64,640],
+                   neural_spline_params=None
+                   ):
+  """Function that returns tomographic density planes 
+  for a given cosmology from a lightcone.
+
+  Args:
+    cosmology: jax-cosmo object
+    density_plane_width: width of the output density slices
+    density_plane_npix: size of the output density slices
+    density_plane_smoothing: Gaussian scale of plane smoothing
+    box_size: [sx,sy,sz] size in Mpc/h of the simulation volume
+    nc: number of particles/voxels in the PM scheme
+    neural_spline_params: optional parameters for neural correction of PM scheme
+  Returns:
+    list of [r, a, plane], slices through the lightcone along with their
+        comoving distance and scale factors.
+  """
+  # Planning out the scale factor stepping to extract desired lensplanes
+  n_lens = int(box_size[-1] // density_plane_width)
+  r = jnp.linspace(0., box_size[-1], n_lens+1)
+  r_center = 0.5*(r[1:] + r[:-1])
+  a_center = jc.background.a_of_chi(cosmology, r_center)
+
+  # Create a small function to generate the matter power spectrum
+  k = jnp.logspace(-4, 1, 128)
```

Comment:
why 128 bins?

Reply:
I've upped to to 256 to be safe. These are points used to interpolate the matter power spectrum, needed when building initial conditions. It shouldn't be very sensitive to this parameter choice.